### ISSUE TITLE
Older unsupported releases are violating the MPL 2.0 due to a broken link

### ISSUE THREAD

**Issue by Jayman2000** [godotengine/godot#98851]
### Tested versions

- Reproducible in commits in the `master` branch that come before d9005912c17ba75e55eca786eb6e46f12a4c85e4.
- Not reproducible in d9005912c17ba75e55eca786eb6e46f12a4c85e4 and commits in the `master` branch that come after that one.

As far as I know, this problem only affects old unsupported versions of Godot. Currently supported versions of Godot are fine. I know that bug reports for old unsupported versions are usually considered invalid, but I think that this bug needs to be an exception for that rule. [We’re still distributing old versions of Godot](https://godotengine.org/download/archive/), so we need to comply with the licenses for all third-party components that are included in those old versions of Godot.

### System information

NixOS 24.05.6122.080166c15633

### Issue description

[We have to comply with version 2.0 of the Mozilla Public License whenever we distribute `thirdparty/certs/ca-certificates.crt`.](https://github.com/godotengine/godot/blob/b00e1cbf743dcb6a2b7f6bd14e348a1a7cf3d403/COPYRIGHT.txt#L182-L185) `ca-certificates.crt` is generated by a tool that converts a `certdata.txt` file into the `ca-certificates.crt` file. In this situation, [`certdata.txt` is the Source Code Form](https://hg.mozilla.org/releases/mozilla-release/raw-file/b8ea2342548b8571e58f9176d9555ccdb5ec199f/security/nss/lib/ckfw/builtins/certdata.txt) and `ca-certificates.crt` is the Executable Form. (The terms “Source Code Form” and “Executable Form” are defined in [section 1 of the MPL 2.0](https://www.mozilla.org/en-US/MPL/2.0/#definitions)).

[Section 3.2 of the MPL 2.0 says (emphasis added):](https://www.mozilla.org/en-US/MPL/2.0/#distribution-of-executable-form)

<blockquote><h3 id="distribution-of-executable-form">3.2. Distribution of Executable Form</h3>
<p>If You distribute Covered Software in Executable Form then:</p>
<ol type="a">
<li><p>such Covered Software must also be made available in Source Code Form, as described in Section&nbsp;3.1, and <b>You must inform recipients of the Executable Form how they can obtain a copy of such Source Code Form</b> by reasonable means in a timely manner, at a charge no more than the cost of distribution to the recipient; and[…]</p></li></ol></blockquote>

This means that whenever we distribute a copy of `ca-certificates.crt`, we have to make sure that we tell people how they can get a copy of the corresponding `certdata.txt` file.

At the moment, we do that by [including a link to the ca-bundle repo in `thirdparty/README.md`](https://github.com/godotengine/godot/tree/b00e1cbf743dcb6a2b7f6bd14e348a1a7cf3d403/thirdparty/README.md#certs). The way that we do it at the moment is totally fine.

---

Some old versions of Godot used a different link. Instead of linking to the ca-bundle repo, [some old versions of Godot include this link](https://github.com/godotengine/godot/blob/688228030850005f84d7b51a527bc779d76b905e/thirdparty/README.md#certs):

> - Upstream: Mozilla, via https://apps.fedoraproject.org/packages/ca-certificates

That link 404s at the moment. The correct new link is <https://packages.fedoraproject.org/pkgs/ca-certificates/ca-certificates/>. This means that we’re accidentally violating the MPL 2.0. We’re distributing copies of `ca-certificates.crt` without telling people how they can get the corresponding `certdata.txt`.

### Steps to reproduce

1. Change directory into a clone of the Godot repo:

    ```bash
    cd <path-to-godot-repo>
    ```

2. Switch to a commit that comes before d9005912c17ba75e55eca786eb6e46f12a4c85e4:

    ```bash
    git switch --detach d9005912c17ba75e55eca786eb6e46f12a4c85e4^
    ```

3. Open `thirdparty/README.md`.

4. Find the “certs” heading.

5. Try opening the link to the Fedora ca-certificates package.

### Minimal reproduction project (MRP)

N/A

**Comment by AThousandShips** [comment]
Where do you find an official release from this time? Does this affect any other branches?

From your link there's no official 4.x releases before January 2022, and the commit you link to was merged in August of 2021

Please link to some official release that has this issue

**Comment by dalexeev** [comment]
> All those branches are EOL so no new release will be made.

I'm not a lawyer, but as far as I understand, the main concern is:

> [We’re still distributing old versions of Godot](https://godotengine.org/download/archive/), so we need to comply with the licenses for all third-party components that are included in those old versions of Godot.

Of course, it's possible there's no copyright at all:

> Also, IANAL, but I would also argue that `certdata.txt` and `ca-certificates.txt` is just plain data (a list of public certificates) which Mozilla has no particular copyright claim on, and thus I question the validity of this MPL 2.0 license claim on this file.

But if we assume the license is violated, is publishing archived unsupported versions considered distribution, or is it considered fair use, due to the archived status? Note that we have now been notified; I am not talking about the period when the link was valid and the period when we were unaware of the potential problem.

**Comment by akien-mga** [comment]
> I'm not a lawyer, but as far as I understand, the main concern is:
> 
> > [We’re still distributing old versions of Godot](https://godotengine.org/download/archive/), so we need to comply with the licenses for all third-party components that are included in those old versions of Godot.

I don't think de-publishing historical releases because a URL expired is a good move for anyone, so I don't intend to act on that.

I'll just add `certdata.txt` to the repo so we always have the source (as was done by bagder following the same request), and commit that to all Git branches we have, supported and EOL.

**Comment by Jayman2000** [comment]
@akien-mga

> There's no way to fix this. We can't change old commits, and we can't force Fedora to reinstate their old link.

I think that there are ways to fix this. The fact that we can’t change old commits makes this problem hard to fix, but not impossible. [For the ca-bundle repository, I overcame the fact that you can’t change old commits by creating a table that contains a list of older commits along with where you can find the corresponding source.](https://github.com/bagder/ca-bundle/pull/12) It was reasonable to do that because the ca-bundle repo only had like 50 commits total. The Godot repo has a lot more commits, so it’s probably not reasonable to do that here.

Perhaps we could add add a statement to [the “Getting the source” page](https://docs.godotengine.org/en/stable/contributing/development/compiling/getting_source.html) that says “some older versions of Godot have some dead links in their source code. Instead of going to <https://apps.fedoraproject.org/packages/ca-certificates>, please go to <https://packages.fedoraproject.org/pkgs/ca-certificates/ca-certificates/>”. Then, we could make sure that pages with older releases link to the “Getting the source” page.

**Comment by akien-mga** [comment]
I don't think this is something worth documenting in current documentation for obsolete releases (Godot 2.1, 3.0, 3.1, and 3.2 specifically).

And we can't fix the tarballs and tags for these old releases so this doesn't solve the "problem". But overall I don't agree that this is a problem, this seems to be a highly theoretical request, which I don't believe comes from an actual need for engine users.

It made sense for bagder to include such information in the readme of a repo's who *sole* purpose is to replicate this data. I don't believe it makes sense for us to do this in our README or documentation, the negative information overload for the vast majority of users who don't care is not worth the added compliance with a license, which, again, I believe to be a bogus claim in the first place.

**Comment by Jayman2000** [comment]
> And we can't fix the tarballs and tags for these old releases so this doesn't solve the "problem".

Why do we need to fix tarballs and tags for old releases in order to solve this problem?

> But overall I don't agree that this is a problem, this seems to be a highly theoretical request, which I don't believe comes from an actual need for engine users.

If it doesn’t come from an actual need for engine users, then where does it come from?

**Comment by akien-mga** [comment]
> Why do we need to fix tarballs and tags for old releases in order to solve this problem?

Because if you download the tarball, or check out the tag, and try to find the source for `ca-certificates.txt`, you'll run into the exact problem you described. We can't change the `thirdparty/README.md` retroactively to fix this.

> If it doesn’t come from an actual need for engine users, then where does it come from?

You tell me :) What's your actual need for this to be made 100% compliant in EOL releases?

Overall, I believe we're doing a very good best effort at compliant with thirdparty licenses, among the best in class from my experience packaging hundreds of open source projects and validating their license compliance.

If Mozilla Legal has a concern, I'm happy to talk about this with them and see how to make sure they're happy with our license compliance. Otherwise, I don't think there's an actual problem here.

**Comment by AThousandShips** [comment]
> I thought that it would be good to help fix this problem for as many people as possible which is why I opened this issue.

But those games won't be helped by anything we do, no one will go to our documentation to check things if they can't open the link of a published game, how could we help in any way with this? And even if we did those games would not be helped, they wouldn't be made compliant by our changes, if the broken link makes them non-compliant *they* need to add similar links themselves. Us having a link where *our* source code is won't fix their executables

**Comment by Faless** [comment]
My 2 cents:

We should ask the copyright holder (Mozilla Foundation) for clarification before we take actions that will likely have no effect on the actual potential breach of such copyright (**assuming that's actually the case** which I'm not convinced of, old games won't be updated, and will stay non-compliant).

**Comment by Jayman2000** [comment]
> But those games won't be helped by anything we do, no one will go to our documentation to check things if they can't open the link of a published game, how could we help in any way with this?

I agree that this change wouldn’t help in that scenario. This won’t automatically fix things if old games linked directly to the Fedora page.

> And even if we did those games would not be helped, they wouldn't be made compliant by our changes, if the broken link makes them non-compliant _they_ need to add similar links themselves. Us having a link where _our_ source code is won't fix their executables

Here was my idea:

1. The Godot project adds “this link is dead, go here instead” documentation somewhere.
2. I add a link to that documentation on those games’s pages

You are correct that it would require manual intervention in order to make the old games compliant, but it would be easier to do the manual intervention this way. If the notice was added to [the “Getting the source” page](https://docs.godotengine.org/en/stable/contributing/development/compiling/getting_source.html), then developers could just make sure that they link to that page regardless of whether the game is using a new or an old version of Godot. At the moment, people have to think about what version of Godot their releases use and figure out if the link in that version is dead.

Basically, it wouldn’t help with the fact that you have to manually add a link, but it would make it easier to figure out what link or links you need to add.

Although, now that I think about it, it is possible that an upstream change would make older games compliant. Most (all?) versions of Godot print this when they start:

```
Godot Engine v4.2.2.stable.nixpkgs.15073afe3 - https://godotengine.org
```

Perhaps adding a notice to the download pages for old versions on godotengine.org  would make some old games automatically compliant because some old games link to godotengine.org. I don’t really know though.

**Comment by adamscott** [comment]
> You are correct that it would require manual intervention in order to make the old games compliant, but it would be easier to do the manual intervention this way.

Again, this would require manual intervention, something that we don't really think is pertinent to do. We are fixing now the issue. And as some said, if the Mozilla Foundation legal team has something to say about this, they will contact the Godot Foundation, and the Foundation will decide of the course of action at that moment.
